src/net/sourceforge/kolmafia/persistence/WitchessSolutionDatabase.java

```diff
@@ -0,0 +1,132 @@
+    }
+  }
+
+  public static class WitchessSolution {
+    protected final int puzzleId;
+    protected final Character[] moves;
+    protected final String coords;
+
+    public WitchessSolution(int puzzleId, Character[] moves, String coords) {
+      this.puzzleId = puzzleId;
+      this.moves = moves;
+      this.coords = coords;
+    }
+
+    public int getId() {
+      return this.puzzleId;
+    }
+
+    public String getCoords() {
+      return this.coords;
+    }
+  }
+
+  private static WitchessSolution parseSolution(Integer puzzleId, String[] data) {
+    String[] solutionSteps = data[1].trim().split(" ");
+    Character[] solution = new Character[solutionSteps.length];
+    for (int i = 0; i < solutionSteps.length; i++) {
+      solution[i] = solutionSteps[i].charAt(0);
+    }
+
+    var coords = solvePath(solution);
+    return new WitchessSolution(puzzleId, solution, coords);
+  }
+
+  private static int[] calculateMidpoint(int[] start, int[] end) {
+    int midX = (start[0] + end[0]) / 2;
+    int midY = (start[1] + end[1]) / 2;
+    return new int[] {midX, midY};
+  }
+
+  private static String solvePath(Character[] moves) {
+    /*
+    We could hard-code the size of each puzzle and therefore the starting coordinate,
+    but to reduce the need for transcribed data, we can instead determine the size of each grid
+    based on the moves made. Since you move from one extreme of the grid to the opposite,
+    we can assume that the max grid size will always be the absolute value of the difference between
+    each move in the cardinal directions (up/down and left/right)
+
+    However, each move increments the grid position by 2, so we need to double the difference
+    to get the real size and coordinates.
+    */
+    Map<Character, Integer> moveCounts = new HashMap<>();
+    for (char move : moves) {
+      moveCounts.put(move, moveCounts.getOrDefault(move, 0) + 1);
+    }
+
+    int netVerticalDisplacement =
+        Math.abs(moveCounts.getOrDefault('u', 0) - moveCounts.getOrDefault('d', 0));
+    int netHorizontalDisplacement =
+        Math.abs(moveCounts.getOrDefault('l', 0) - moveCounts.getOrDefault('r', 0));
+
+    int maxX = netVerticalDisplacement * 2;
+    int maxY = netHorizontalDisplacement * 2;
+
+    int x = maxX, y = 0;
+    Set<String> path = new HashSet<>();
+
+    for (char move : moves) {
+      int[] start = new int[] {x, y};
+      int[] delta = moveDict.get(move);
+      x = Math.max(0, Math.min(x + delta[0], maxX));
+      y = Math.max(0, Math.min(y + delta[1], maxY));
+      int[] end = new int[] {x, y};
+      int[] midpoint = calculateMidpoint(start, end);
+      path.add(midpoint[0] + "," + midpoint[1]);
+    }
+
+    List<String> sortedPath = new ArrayList<>(path);
+    sortedPath.sort(
+        Comparator.comparingInt((String s) -> Integer.parseInt(s.split(",")[0]))
+            .thenComparing(s -> Integer.parseInt(s.split(",")[1])));
+
+    return String.join("|", sortedPath);
```

Comment:
it would be nice to have a test that gets the solution for a certain puzzle and confirms this generated string is as expected.

It would separately be nice to have some documentation at the top of this function as to what this generated string is supposed to contain, and how that relates to the puzzle itself.

Reply:
I do have a txt file locally with the final GET request strings that I was planning to write a test with. I wasn't sure the best approach to reading from that .txt file without writing another Database class to do so, though, so opted to exclude it originally.

Reply:
you don't need all of them -- one would be enough, I think.

If you did want all of them (or many), I would use a `ParameterizedTest` with a `CsvSource` for the puzzle number and the expected string.

Reply:
I've got 15 of them saved, I might grab some from my puzzles tomorrow since the ones I currently have are in the earlier ranges, and I think it'd be good to have later puzzles as well since the board can get oddly shaped. I'll pick probably 5 from a few different spots in the order.

Reply:
Alright, I created WitchessSolutionDatabaseTest.java and check that the generated URL for 10 puzzles match the network requests I grabbed when manually solving the puzzles.